src/java/main/org/apache/zookeeper/server/ZooKeeperServerMain.java

```diff
@@ -146,14 +146,26 @@ public void runFromConfig(ServerConfig config) throws IOException, AdminServerEx
             );
             containerManager.start();
 
+            // Watch status of ZooKeeper server. If there is an internal error
+            // then will do a graceful shutdown.
+            while (zkServer.isRunning()) {
```

Comment:
For quorum it was easy because ZK has Leader/Follower/Observer thread to read the packets and has an existing polling mechanism. But for a standalone server we don't have such polling mechanism and I think need to introduce some mechanism to listen ZKServer state changes and notify the main thread. Earlier we discussed an approach by adding an extra HealthMonitorThread but decided to move that logic inside main thread - https://issues.apache.org/jira/browse/ZOOKEEPER-2247?focusedCommentId=15126934&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15126934

Thanks for the latch suggestion. But I think here also we should have a separate notifier thread to poll the server state changes and countdown latch, right?